compiler/src/iree/compiler/GlobalOptimization/RaiseSpecialOps.cpp

```diff
@@ -990,6 +990,170 @@ class ConcatenateNegateAndSlicePattern
   }
 };
 
+//===----------------------------------------------------------------------===//
+// Attention -> Transpose fusion
+//===----------------------------------------------------------------------===//
+
+/// Pattern to fold
+///
+/// ```mlir
+/// %0 = iree_linalg_ext.attention {
+///     indexing_maps = [affine_map<(d0, d1, d2, d3, d4) -> (d0, d1, d2)>,
+///                      affine_map<(d0, d1, d2, d3, d4) -> (d0, d3, d2)>,
+///                      affine_map<(d0, d1, d2, d3, d4) -> (d0, d3, d4)>,
+///                      affine_map<(d0, d1, d2, d3, d4) -> (d0, d1, d4)>]}
+///     ins(%query, %key, %value) ....
+/// %1 = tensor.expand_shape %0 into [[0, 1], [2], [3]] ....
+/// %2 = linalg.generic {
+///     indexing_maps = [affine_map<(d0, d1, d2, d3) -> (d0, d2, d1, d3)>,
+///                      affine_map<(d0, d1, d2, d3) -> (d0, d1, d2, d3)>]}
+///     ins(%1)
+/// ```
+///
+/// to
+///
+/// ```
+/// %0 = iree_linalg_ext.attention {
+///     indexing_maps = [affine_map<(d0, d00, d1, d2, d3, d4) -> (d0, d00, d1,
+///     d2)>,
+///                      affine_map<(d0, d00, d1, d2, d3, d4) -> (d0, d00, d3,
+///                      d2)>, affine_map<(d0, d00, d1, d2, d3, d4) -> (d0, d00,
+///                      d3, d4)>, affine_map<(d0, d00, d1, d2, d3, d4) -> (d0,
+///                      d1, d00, d4)>]}
+///     ins(%expanded_query, %expanded_key, %expanded_value) ....
+/// ```
+///
+///  Do a very specific match for now. Eventually this can be generalized to a
+///  use similar analysis as to what the reshape propagation across Linalg op
+///  does. TODO(#17673)
+///
+struct FoldAttentionAndTranspose
+    : public OpRewritePattern<IREE::LinalgExt::AttentionOp> {
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(IREE::LinalgExt::AttentionOp attentionOp,
+                                PatternRewriter &rewriter) const override {
+    // Check for single use attention op.
+    if (!attentionOp->hasOneUse()) {
+      return rewriter.notifyMatchFailure(attentionOp,
+                                         "attention op has multiple uses");
+    }
+    auto expandShapeOp =
+        dyn_cast<tensor::ExpandShapeOp>(*attentionOp->user_begin());
+    if (!expandShapeOp) {
+      return rewriter.notifyMatchFailure(attentionOp,
+                                         "user is not an expand shape op.");
+    }
+    // Check for single use of expand shape op.
+    if (!expandShapeOp->hasOneUse()) {
+      return rewriter.notifyMatchFailure(attentionOp,
+                                         "expand shape op has multiple uses");
+    }
+    auto transposeLikeOp =
+        dyn_cast<linalg::LinalgOp>(*expandShapeOp->user_begin());
+    if (!transposeLikeOp) {
+      return failure();
+    }
+    if (!(transposeLikeOp.getNumDpsInputs() == 1 &&
```

Comment:
[No action required] I'm surprised, we haven't got a helper function for the transposeLikeOp either. I know we have `LinalgOpInfo(op).isTranspose()` but I don't think it's doing this haha.